tests/suites/test_suite_aes.function

```diff
@@ -151,6 +151,157 @@ exit:
+    unsigned char *data_unit;
+    unsigned char *src;
+    unsigned char *dst;
+    unsigned char *output;
+    mbedtls_aes_xts_context ctx;
+    size_t key_len, src_len, dst_len, data_unit_len;
+
+    mbedtls_aes_xts_init( &ctx );
+
+    data_unit = unhexify_alloc( hex_data_unit_string, &data_unit_len );
+    TEST_ASSERT( data_unit_len == AES_BLOCK_SIZE );
+
+    key = unhexify_alloc( hex_key_string, &key_len );
+    TEST_ASSERT( key_len % 2 == 0 );
+
+    src = unhexify_alloc( hex_src_string, &src_len );
+    dst = unhexify_alloc( hex_dst_string, &dst_len );
+    TEST_ASSERT( src_len == dst_len );
+
+    output = zero_alloc(dst_len);
+
+    mbedtls_aes_xts_setkey_enc( &ctx, key, key_len * 8 );
+    TEST_ASSERT( mbedtls_aes_crypt_xts( &ctx, MBEDTLS_AES_ENCRYPT, src_len, data_unit, src, output ) == 0 ); // XXX Code Review: must this be one line?
+
+    TEST_ASSERT( memcmp( output, dst, dst_len ) == 0 );
+
+exit:
+    mbedtls_aes_xts_free( &ctx );
+}
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_CIPHER_MODE_XTS */
+void aes_decrypt_xts( char *hex_key_string, char *hex_data_unit_string,
+                      char *hex_dst_string, char *hex_src_string )
+{
+    /* XXX Code Review: I don't really like all this duplicated code between
+     * encrypt and decrypt. The only thing different is the mode and key
+     * initialization. */
+    enum { AES_BLOCK_SIZE = 16 };
+    unsigned char *key;
+    unsigned char *data_unit;
+    unsigned char *src;
+    unsigned char *dst;
+    unsigned char *output;
+    mbedtls_aes_xts_context ctx;
+    size_t key_len, src_len, dst_len, data_unit_len;
+
+    mbedtls_aes_xts_init( &ctx );
+
+    data_unit = unhexify_alloc( hex_data_unit_string, &data_unit_len );
+    TEST_ASSERT( data_unit_len == AES_BLOCK_SIZE );
+
+    key = unhexify_alloc( hex_key_string, &key_len );
+    TEST_ASSERT( key_len % 2 == 0 );
+    src = unhexify_alloc( hex_src_string, &src_len );
+    dst = unhexify_alloc( hex_dst_string, &dst_len );
+    TEST_ASSERT( src_len == dst_len );
+
+    output = zero_alloc(dst_len);
+
+    mbedtls_aes_xts_setkey_dec( &ctx, key, key_len * 8 );
+    TEST_ASSERT( mbedtls_aes_crypt_xts( &ctx, MBEDTLS_AES_DECRYPT, src_len, data_unit, src, output ) == 0 ); // XXX Code Review: must this be one line?
+
+    TEST_ASSERT( memcmp( output, dst, dst_len ) == 0 );
+
+exit:
+    mbedtls_aes_xts_free( &ctx );
+}
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_CIPHER_MODE_XTS */
+void aes_crypt_xts_size( int size, int retval )
+{
+    size_t length = size;
+    /* Note that this function will most likely crash on failure, as NULL
+     * parameters will be used. In the passing case, the length check in
+     * mbedtls_aes_crypt_xts() will prevent any accesses to parameters by
+     * exiting the function early. XXX <-- Hey look, Mr. Codereviewer. What do
+     * you think? Any ideas how to be safe and not have to actually allocate a
+     * 16 MiB buffer nor add NULL checks to mbedtls_aes_crypt_xts()? */
```

Comment:
I'm fine with the risk of crashing if the length check is not implemented correctly. This is C, it happens sometimes that broken code causes the tests to crash.

On the other hand, at least the context and index should be valid, so that we don't get either a crash because the code tries to dereference them before checking `length` (at least for the context, that's plausible) or a false positive because the code checks those parameters for nullhood before checking `length`.

Due to the same risk of a false positive if the code has additional null checks before the length check, `input` and `output` should be non-null. I propose to make them pointers to a single byte. This makes the behavior on a failed test undefined, but at least there's less of a risk that the test will pass spuriously.

Reply:
I've added an initialized context to the test, but did not set any keys in it.

As for input and output being NULL, I would expect a different error returned for NULL checks than for input length. The test is checking for this. I'd rather a NULL get deferenced than some funny values from the stack in the case of an overread.